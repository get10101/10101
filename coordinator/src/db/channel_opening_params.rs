use crate::schema::channel_opening_params;
use bitcoin::Amount;
use diesel::ExpressionMethods;
use diesel::Insertable;
use diesel::OptionalExtension;
use diesel::PgConnection;
use diesel::QueryDsl;
use diesel::QueryResult;
use diesel::Queryable;
use diesel::QueryableByName;
use diesel::RunQueryDsl;
use time::OffsetDateTime;
use uuid::Uuid;

#[derive(Queryable, QueryableByName, Insertable, Debug, Clone, PartialEq)]
#[diesel(table_name = channel_opening_params)]
pub struct ChannelOpeningParams {
    order_id: String,
    coordinator_reserve: i64,
    trader_reserve: i64,
    created_at: i64,
    external_funding: Option<i64>,
}

pub fn insert(
    conn: &mut PgConnection,
    order_id: Uuid,
    channel_opening_params: crate::ChannelOpeningParams,
) -> QueryResult<()> {
    let affected_rows = diesel::insert_into(channel_opening_params::table)
        .values(ChannelOpeningParams::from((
            order_id,
            channel_opening_params,
        )))
        .execute(conn)?;

    if affected_rows == 0 {
        return diesel::result::QueryResult::Err(diesel::result::Error::NotFound);
    }

    diesel::result::QueryResult::Ok(())
}

pub fn get_by_order_id(
    conn: &mut PgConnection,
    order_id: Uuid,
) -> QueryResult<Option<crate::ChannelOpeningParams>> {
    let channel_opening_params: Option<ChannelOpeningParams> = channel_opening_params::table
        .filter(channel_opening_params::order_id.eq(order_id.to_string()))
        .first(conn)
        .optional()?;

    Ok(channel_opening_params.map(crate::ChannelOpeningParams::from))
}

impl From<(Uuid, crate::ChannelOpeningParams)> for ChannelOpeningParams {
    fn from((order_id, channel_opening_params): (Uuid, crate::ChannelOpeningParams)) -> Self {
        Self {
            order_id: order_id.to_string(),
            trader_reserve: channel_opening_params.trader_reserve.to_sat() as i64,
            coordinator_reserve: channel_opening_params.coordinator_reserve.to_sat() as i64,
            external_funding: channel_opening_params
                .external_funding
                .map(|funding| funding.to_sat() as i64),
            created_at: OffsetDateTime::now_utc().unix_timestamp(),
        }
    }
}

impl From<ChannelOpeningParams> for crate::ChannelOpeningParams {
    fn from(value: ChannelOpeningParams) -> Self {
        Self {
            coordinator_reserve: Amount::from_sat(value.coordinator_reserve as u64),
            trader_reserve: Amount::from_sat(value.trader_reserve as u64),
            external_funding: value
                .external_funding
                .map(|funding| Amount::from_sat(funding as u64)),
        }
    }
}
