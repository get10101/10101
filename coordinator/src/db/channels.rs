use crate::schema;
use crate::schema::channels;
use crate::schema::sql_types::ChannelStateType;
use anyhow::ensure;
use anyhow::Context;
use anyhow::Result;
use bitcoin::hashes::hex::ToHex;
use bitcoin::secp256k1::PublicKey;
use bitcoin::Txid;
use diesel::query_builder::QueryId;
use diesel::AsChangeset;
use diesel::AsExpression;
use diesel::BoolExpressionMethods;
use diesel::ExpressionMethods;
use diesel::FromSqlRow;
use diesel::Insertable;
use diesel::OptionalExtension;
use diesel::PgConnection;
use diesel::QueryDsl;
use diesel::QueryResult;
use diesel::Queryable;
use diesel::QueryableByName;
use diesel::RunQueryDsl;
use dlc_manager::ChannelId;
use hex::FromHex;
use lightning::ln::PaymentHash;
use ln_dlc_node::channel::UserChannelId;
use std::any::TypeId;
use std::str::FromStr;
use time::OffsetDateTime;

#[derive(Debug, Clone, Copy, PartialEq, FromSqlRow, AsExpression)]
#[diesel(sql_type = ChannelStateType)]
pub(crate) enum ChannelState {
    Announced,
    Pending,
    Open,
    Closed,
    ForceClosedRemote,
    ForceClosedLocal,
}

impl QueryId for ChannelStateType {
    type QueryId = ChannelStateType;
    const HAS_STATIC_QUERY_ID: bool = false;

    fn query_id() -> Option<TypeId> {
        None
    }
}

#[derive(Insertable, QueryableByName, Queryable, Debug, Clone, PartialEq, AsChangeset)]
#[diesel(table_name = channels)]
pub(crate) struct Channel {
    pub user_channel_id: String,
    pub channel_id: Option<String>,
    pub inbound_sats: i64,
    pub outbound_sats: i64,
    pub funding_txid: Option<String>,
    pub channel_state: ChannelState,
    pub counterparty_pubkey: String,
    pub created_at: OffsetDateTime,
    pub updated_at: OffsetDateTime,
    pub open_channel_fee_payment_hash: Option<String>,
    pub liquidity_option_id: Option<i32>,
}

pub(crate) fn get(user_channel_id: &str, conn: &mut PgConnection) -> QueryResult<Option<Channel>> {
    channels::table
        .filter(channels::user_channel_id.eq(user_channel_id))
        .first(conn)
        .optional()
}

pub(crate) fn get_announced_channel(
    counterparty_pubkey: &str,
    conn: &mut PgConnection,
) -> QueryResult<Option<Channel>> {
    channels::table
        .filter(channels::counterparty_pubkey.eq(counterparty_pubkey))
        .filter(channels::channel_state.eq(ChannelState::Announced))
        .first(conn)
        .optional()
}

pub(crate) fn get_all_non_pending_channels(conn: &mut PgConnection) -> QueryResult<Vec<Channel>> {
    channels::table
        .filter(
            channels::channel_state
                .ne(ChannelState::Pending)
                .and(schema::channels::funding_txid.is_not_null()),
        )
        .load(conn)
}

pub(crate) fn update_payment_hash(
    payment_hash: PaymentHash,
    funding_txid: String,
    conn: &mut PgConnection,
) -> Result<()> {
    let mut channel: Channel = channels::table
        .filter(channels::funding_txid.eq(funding_txid.clone()))
        .first(conn)
        .with_context(|| format!("No channel found for funding txid {funding_txid}"))?;

    channel.open_channel_fee_payment_hash = Some(payment_hash.0.to_hex());
    upsert(channel, conn)
}

pub fn get_by_channel_id(
    channel_id: String,
    conn: &mut PgConnection,
) -> Result<Option<ln_dlc_node::channel::Channel>> {
    let channel = channels::table
        .filter(channels::channel_id.eq(channel_id))
        .first::<Channel>(conn)
        .optional()?
        .map(ln_dlc_node::channel::Channel::from);
    Ok(channel)
}

pub(crate) fn upsert(channel: Channel, conn: &mut PgConnection) -> Result<()> {
    let affected_rows = diesel::insert_into(channels::table)
        .values(channel.clone())
        .on_conflict(schema::channels::user_channel_id)
        .do_update()
        .set(&channel)
        .execute(conn)?;

    ensure!(affected_rows > 0, "Could not upsert channel");

    Ok(())
}

impl From<ln_dlc_node::channel::Channel> for Channel {
    fn from(value: ln_dlc_node::channel::Channel) -> Self {
        Channel {
            user_channel_id: value.user_channel_id.to_string(),
            channel_id: value.channel_id.map(|cid| cid.to_hex()),
            liquidity_option_id: value.liquidity_option_id,
            inbound_sats: value.inbound_sats as i64,
            outbound_sats: value.outbound_sats as i64,
            funding_txid: value.funding_txid.map(|txid| txid.to_string()),
            channel_state: value.channel_state.into(),
            counterparty_pubkey: value.counterparty.to_string(),
            created_at: value.created_at,
            updated_at: value.updated_at,
            open_channel_fee_payment_hash: None,
        }
    }
}

impl From<ln_dlc_node::channel::ChannelState> for ChannelState {
    fn from(value: ln_dlc_node::channel::ChannelState) -> Self {
        match value {
            ln_dlc_node::channel::ChannelState::Announced => ChannelState::Announced,
            ln_dlc_node::channel::ChannelState::Pending => ChannelState::Pending,
            ln_dlc_node::channel::ChannelState::Open => ChannelState::Open,
            // mapping `ChannelState::OpenUnpaid` to open as the coordinator references the payment
            // through the payment hash. Hence no need to reflect that state twice.
            ln_dlc_node::channel::ChannelState::OpenUnpaid => ChannelState::Open,
            ln_dlc_node::channel::ChannelState::Closed => ChannelState::Closed,
            ln_dlc_node::channel::ChannelState::ForceClosedLocal => ChannelState::ForceClosedLocal,
            ln_dlc_node::channel::ChannelState::ForceClosedRemote => {
                ChannelState::ForceClosedRemote
            }
        }
    }
}

impl From<Channel> for ln_dlc_node::channel::Channel {
    fn from(value: Channel) -> Self {
        ln_dlc_node::channel::Channel {
            user_channel_id: UserChannelId::try_from(value.user_channel_id)
                .expect("valid user channel id"),
            channel_id: value
                .channel_id
                .map(|cid| ChannelId::from_hex(cid).expect("valid channel id")),
            liquidity_option_id: value.liquidity_option_id,
            inbound_sats: value.inbound_sats as u64,
            outbound_sats: value.outbound_sats as u64,
            funding_txid: value
                .funding_txid
                .map(|txid| Txid::from_str(&txid).expect("valid txid")),
            channel_state: value.channel_state.into(),
            counterparty: PublicKey::from_str(&value.counterparty_pubkey)
                .expect("valid public key"),
            created_at: value.created_at,
            updated_at: value.updated_at,
        }
    }
}

impl From<ChannelState> for ln_dlc_node::channel::ChannelState {
    fn from(value: ChannelState) -> Self {
        match value {
            ChannelState::Announced => ln_dlc_node::channel::ChannelState::Announced,
            ChannelState::Pending => ln_dlc_node::channel::ChannelState::Pending,
            ChannelState::Open => ln_dlc_node::channel::ChannelState::Open,
            ChannelState::Closed => ln_dlc_node::channel::ChannelState::Closed,
            ChannelState::ForceClosedLocal => ln_dlc_node::channel::ChannelState::ForceClosedLocal,
            ChannelState::ForceClosedRemote => {
                ln_dlc_node::channel::ChannelState::ForceClosedRemote
            }
        }
    }
}
