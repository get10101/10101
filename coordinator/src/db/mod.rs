pub mod bonus_status;
pub mod bonus_tiers;
pub mod channel_opening_params;
pub mod collaborative_reverts;
pub mod custom_types;
pub mod dlc_channels;
pub mod dlc_messages;
pub mod dlc_protocols;
pub mod last_outbound_dlc_message;
pub mod liquidity;
pub mod liquidity_options;
pub mod orders_helper;
pub mod polls;
pub mod positions;
pub mod spendable_outputs;
pub mod trade_params;
pub mod trades;
pub mod transactions;
pub mod user;
