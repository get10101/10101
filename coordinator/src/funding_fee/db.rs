pub mod funding_fee_events;
pub mod funding_rates;
pub mod protocol_funding_fee_events;

pub use funding_fee_events::*;
pub use funding_rates::*;
pub use protocol_funding_fee_events::*;
