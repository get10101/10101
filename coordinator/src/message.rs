use crate::notifications::Notification;
use crate::notifications::NotificationKind;
use anyhow::Context;
use anyhow::Result;
use bitcoin::secp256k1::PublicKey;
use futures::future::RemoteHandle;
use futures::FutureExt;
use parking_lot::RwLock;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::broadcast;
use tokio::sync::broadcast::error::RecvError;
use tokio::sync::mpsc;
use tokio::sync::mpsc::Sender;
use xxi_node::commons::Message;

/// This value is arbitrarily set to 100 and defines the message accepted in the message
/// channel buffer.
const NOTIFICATION_BUFFER_SIZE: usize = 100;

/// Message sent to users via the websocket.
#[derive(Debug)]
pub struct TraderMessage {
    pub trader_id: PublicKey,
    pub message: Message,
    pub notification: Option<NotificationKind>,
}

#[derive(Clone)]
pub struct NewUserMessage {
    pub new_user: PublicKey,
    pub sender: Sender<Message>,
}

pub fn spawn_delivering_messages_to_authenticated_users(
    notification_sender: Sender<Notification>,
    tx_user_feed: broadcast::Sender<NewUserMessage>,
) -> (RemoteHandle<()>, Sender<TraderMessage>) {
    let (sender, mut receiver) = mpsc::channel::<TraderMessage>(NOTIFICATION_BUFFER_SIZE);

    let authenticated_users = Arc::new(RwLock::new(HashMap::new()));

    tokio::task::spawn({
        let traders = authenticated_users.clone();
        async move {
            let mut user_feed = tx_user_feed.subscribe();
            loop {
                match user_feed.recv().await {
                    Ok(new_user_msg) => {
                        traders
                            .write()
                            .insert(new_user_msg.new_user, new_user_msg.sender);
                    }
                    Err(RecvError::Closed) => {
                        tracing::error!("New user message sender died! Channel closed");
                        break;
                    }
                    Err(RecvError::Lagged(skip)) => {
                        tracing::warn!(%skip, "Lagging behind on new user message")
                    }
                }
            }
        }
    });

    let (fut, remote_handle) = {
        async move {
            while let Some(trader_message) = receiver.recv().await {
                if let Err(e) = process_trader_message(
                    &authenticated_users,
                    &notification_sender,
                    trader_message,
                )
                .await
                {
                    tracing::error!("Failed to process trader message: {e:#}");
                }
            }

            tracing::error!("Channel closed");
        }
        .remote_handle()
    };

    tokio::spawn(fut);

    (remote_handle, sender)
}

async fn process_trader_message(
    authenticated_users: &RwLock<HashMap<PublicKey, Sender<Message>>>,
    notification_sender: &Sender<Notification>,
    trader_message: TraderMessage,
) -> Result<()> {
    let trader_id = trader_message.trader_id;
    let message = trader_message.message;
    let notification = trader_message.notification;
    tracing::info!(%trader_id, ?message, "Sending trader message");

    let trader = authenticated_users.read().get(&trader_id).cloned();

    match trader {
        Some(sender) => {
            if let Err(e) = sender.send(message).await {
                tracing::warn!(%trader_id, "Connection lost to trader: {e:#}");
            } else {
                tracing::trace!(
                    %trader_id,
                    "Skipping optional push notifications as the user was successfully \
                     notified via the websocket"
                );
                return Ok(());
            }
        }
        None => tracing::warn!(%trader_id, "Trader is not connected"),
    };

    if let Some(notification_kind) = notification {
        tracing::debug!(%trader_id, "Sending push notification to user");

        notification_sender
            .send(Notification::new(trader_id, notification_kind))
            .await
            .with_context(|| format!("Failed to send push notification to trader {trader_id}"))?;
    }

    Ok(())
}
