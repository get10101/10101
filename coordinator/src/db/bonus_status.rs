use crate::db::bonus_tiers;
use crate::schema::bonus_status;
use crate::schema::sql_types::BonusStatusType;
use bitcoin::secp256k1::PublicKey;
use diesel::AsExpression;
use diesel::ExpressionMethods;
use diesel::FromSqlRow;
use diesel::Insertable;
use diesel::PgConnection;
use diesel::QueryDsl;
use diesel::QueryResult;
use diesel::Queryable;
use diesel::RunQueryDsl;
use time::OffsetDateTime;

/// A user's referral bonus status may be active for this much days at max
const MAX_DAYS_FOR_ACTIVE_REFERRAL_STATUS: i64 = 30;

#[derive(Debug, Clone, Copy, PartialEq, FromSqlRow, AsExpression)]
#[diesel(sql_type = BonusStatusType)]
pub enum BonusType {
    /// The bonus is because he referred so many users
    Referral,
    /// The user has been referred and gets a bonus
    Referent,
}

#[allow(dead_code)]
// this is needed because the fields needs to be here to satisfy diesel
#[derive(Queryable, Debug, Clone)]
#[diesel(table_name = bonus_status)]
pub(crate) struct BonusStatus {
    pub(crate) id: i32,
    pub(crate) trader_pubkey: String,
    pub(crate) tier_level: i32,
    pub(crate) fee_rebate: f32,
    pub(crate) bonus_type: BonusType,
    pub(crate) activation_timestamp: OffsetDateTime,
    pub(crate) deactivation_timestamp: OffsetDateTime,
}

impl From<BonusType> for commons::BonusStatusType {
    fn from(value: BonusType) -> Self {
        match value {
            BonusType::Referral => commons::BonusStatusType::Referral,
            BonusType::Referent => commons::BonusStatusType::Referent,
        }
    }
}

#[derive(Insertable, Debug, Clone)]
#[diesel(table_name = bonus_status)]
pub(crate) struct NewBonusStatus {
    pub(crate) trader_pubkey: String,
    pub(crate) tier_level: i32,
    pub(crate) fee_rebate: f32,
    pub(crate) bonus_type: BonusType,
    pub(crate) activation_timestamp: OffsetDateTime,
    pub(crate) deactivation_timestamp: OffsetDateTime,
}

/// This function might return multiple status for a single user
///
/// Because he might have moved up into the next level without the old level being expired. The
/// caller is responsible in picking the most suitable status
pub(crate) fn active_status_for_user(
    conn: &mut PgConnection,
    trader_pubkey: &PublicKey,
) -> QueryResult<Vec<BonusStatus>> {
    bonus_status::table
        .filter(bonus_status::trader_pubkey.eq(trader_pubkey.to_string()))
        .filter(bonus_status::deactivation_timestamp.gt(OffsetDateTime::now_utc()))
        .load(conn)
}

pub(crate) fn insert(
    conn: &mut PgConnection,
    trader_pk: &PublicKey,
    tier_level: i32,
    bonus_type: BonusType,
) -> QueryResult<BonusStatus> {
    let existing_status_for_user = active_status_for_user(conn, trader_pk)?;
    let bonus_tier = bonus_tiers::tier_by_tier_level(conn, tier_level)?;

    if let Some(status) = existing_status_for_user
        .into_iter()
        .find(|status| status.tier_level == tier_level)
    {
        tracing::debug!(
            trader_pubkey = trader_pk.to_string(),
            tier_level,
            "User has already gained bonus status"
        );
        return Ok(status);
    }

    let bonus_status = diesel::insert_into(bonus_status::table)
        .values(NewBonusStatus {
            trader_pubkey: trader_pk.to_string(),
            tier_level,
            fee_rebate: bonus_tier.fee_rebate,
            bonus_type,
            activation_timestamp: OffsetDateTime::now_utc(),
            deactivation_timestamp: OffsetDateTime::now_utc()
                + time::Duration::days(MAX_DAYS_FOR_ACTIVE_REFERRAL_STATUS),
        })
        .get_result(conn)?;

    Ok(bonus_status)
}
