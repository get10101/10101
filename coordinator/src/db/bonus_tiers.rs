use crate::db::bonus_status::BonusType;
use crate::schema::bonus_tiers;
use bitcoin::secp256k1::PublicKey;
use diesel::pg::sql_types::Timestamptz;
use diesel::prelude::*;
use diesel::sql_query;
use diesel::sql_types::Float;
use diesel::sql_types::Text;
use diesel::PgConnection;
use diesel::QueryResult;
use diesel::Queryable;
use rust_decimal::Decimal;
use time::OffsetDateTime;

pub struct Referral {
    pub volume: Decimal,
}

#[derive(Queryable, Debug, Clone)]
#[diesel(table_name = bonus_tiers)]
// this is needed because some fields are unused but need to be here for diesel
#[allow(dead_code)]
pub(crate) struct BonusTier {
    pub(crate) id: i32,
    pub(crate) tier_level: i32,
    pub(crate) min_users_to_refer: i32,
    pub(crate) fee_rebate: f32,
    pub(crate) bonus_tier_type: BonusType,
    pub(crate) active: bool,
}

/// Returns all active bonus tiers for given types
pub(crate) fn all_active_by_type(
    conn: &mut PgConnection,
    types: Vec<BonusType>,
) -> QueryResult<Vec<BonusTier>> {
    bonus_tiers::table
        .filter(bonus_tiers::active.eq(true))
        .filter(bonus_tiers::bonus_tier_type.eq_any(types))
        .load::<BonusTier>(conn)
}

pub(crate) fn tier_by_tier_level(
    conn: &mut PgConnection,
    tier_level: i32,
) -> QueryResult<BonusTier> {
    bonus_tiers::table
        .filter(bonus_tiers::tier_level.eq(tier_level))
        .first(conn)
}

#[derive(Debug, QueryableByName, Clone)]
pub struct UserReferralSummaryView {
    #[diesel(sql_type = Text)]
    pub referring_user: String,
    #[diesel(sql_type = Text)]
    pub referring_user_referral_code: String,
    #[diesel(sql_type = Text)]
    pub referred_user: String,
    #[diesel(sql_type = Text)]
    pub referred_user_referral_code: String,
    #[diesel(sql_type = Timestamptz)]
    pub timestamp: OffsetDateTime,
    #[diesel(sql_type = Float)]
    pub referred_user_total_quantity: f32,
}

/// Returns all referred users for by referrer with trading volume > 0
pub(crate) fn all_referrals_by_referring_user(
    conn: &mut PgConnection,
    trader_pubkey: &PublicKey,
) -> QueryResult<Vec<UserReferralSummaryView>> {
    // we have to do this manually because diesel does not support views. If you make a change to
    // below, make sure to test this against a life db as errors will only be thrown at runtime
    let query = "SELECT referring_user, referring_user_referral_code, \
                referred_user, \
                referred_user_referral_code, \
                timestamp, \
                referred_user_total_quantity \
                    FROM user_referral_summary_view where referring_user = $1 \
                    and referred_user_total_quantity > 0";
    let summaries: Vec<UserReferralSummaryView> = sql_query(query)
        .bind::<Text, _>(trader_pubkey.to_string())
        .load(conn)?;

    Ok(summaries)
}
