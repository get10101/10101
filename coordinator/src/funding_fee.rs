use crate::db;
use crate::decimal_from_f32;
use crate::message::OrderbookMessage;
use crate::to_nearest_hour_in_the_past;
use anyhow::bail;
use anyhow::Context;
use anyhow::Result;
use bitcoin::secp256k1::PublicKey;
use bitcoin::SignedAmount;
use diesel::r2d2::ConnectionManager;
use diesel::r2d2::Pool;
use diesel::PgConnection;
use rust_decimal::prelude::ToPrimitive;
use rust_decimal::Decimal;
use rust_decimal::RoundingStrategy;
use rust_decimal_macros::dec;
use std::time::Duration;
use time::ext::NumericalDuration;
use time::format_description;
use time::OffsetDateTime;
use tokio::task::block_in_place;
use tokio_cron_scheduler::JobScheduler;
use xxi_node::commons::ContractSymbol;
use xxi_node::commons::Direction;
use xxi_node::commons::Message;

const RETRY_INTERVAL: Duration = Duration::from_secs(5);

/// The funding rate for any position opened before the `end_date`, which remained open through the
/// `end_date`.
#[derive(Clone, Debug)]
pub struct FundingRate {
    /// A positive funding rate indicates that longs pay shorts; a negative funding rate indicates
    /// that shorts pay longs.
    rate: Decimal,
    /// The start date for the funding rate period. This value is only used for informational
    /// purposes.
    ///
    /// The `start_date` is always a whole hour.
    start_date: OffsetDateTime,
    /// The end date for the funding rate period. When the end date has passed, all active
    /// positions that were created before the end date should be charged a funding fee based
    /// on the `rate`.
    ///
    /// The `end_date` is always a whole hour.
    end_date: OffsetDateTime,
}

impl FundingRate {
    pub(crate) fn new(rate: Decimal, start_date: OffsetDateTime, end_date: OffsetDateTime) -> Self {
        let start_date = to_nearest_hour_in_the_past(start_date);
        let end_date = to_nearest_hour_in_the_past(end_date);

        Self {
            rate,
            start_date,
            end_date,
        }
    }

    pub fn rate(&self) -> Decimal {
        self.rate
    }

    pub fn start_date(&self) -> OffsetDateTime {
        self.start_date
    }

    pub fn end_date(&self) -> OffsetDateTime {
        self.end_date
    }
}

/// A record that a funding fee is owed between the coordinator and a trader.
#[derive(Clone, Copy, Debug)]
pub struct FundingFeeEvent {
    pub id: i32,
    /// A positive amount indicates that the trader pays the coordinator; a negative amount
    /// indicates that the coordinator pays the trader.
    pub amount: SignedAmount,
    pub trader_pubkey: PublicKey,
    pub position_id: i32,
    pub due_date: OffsetDateTime,
    pub price: Decimal,
    pub funding_rate: Decimal,
    pub paid_date: Option<OffsetDateTime>,
}

impl From<FundingFeeEvent> for xxi_node::message_handler::FundingFeeEvent {
    fn from(value: FundingFeeEvent) -> Self {
        Self {
            due_date: value.due_date,
            funding_rate: value.funding_rate,
            price: value.price,
            funding_fee: value.amount,
        }
    }
}

#[derive(Clone, Copy, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
pub enum IndexPriceSource {
    Bitmex,
    /// The index price will be hard-coded for testing.
    Test,
}

pub async fn generate_funding_fee_events_periodically(
    scheduler: &JobScheduler,
    pool: Pool<ConnectionManager<PgConnection>>,
    auth_users_notifier: tokio::sync::mpsc::Sender<OrderbookMessage>,
    schedule: String,
    index_price_source: IndexPriceSource,
) -> Result<()> {
    scheduler
        .add(tokio_cron_scheduler::Job::new(
            schedule.as_str(),
            move |_, _| {
                let mut attempts_left = 10;

                // We want to retry
                while let (Err(e), true) = (
                    generate_funding_fee_events(
                        &pool,
                        index_price_source,
                        auth_users_notifier.clone(),
                    ),
                    attempts_left > 0,
                ) {
                    attempts_left -= 1;

                    tracing::error!(
                        retry_interval = ?RETRY_INTERVAL,
                        attempts_left,
                        "Failed to generate funding fee events: {e:#}. \
                         Trying again"
                    );

                    std::thread::sleep(RETRY_INTERVAL);
                }
            },
        )?)
        .await?;

    scheduler.start().await?;

    Ok(())
}

/// Generate [`FundingFeeEvent`]s for all active positions.
///
/// When called, a [`FundingFeeEvent`] will be generated for an active position if:
///
/// - We can get a [`FundingRate`] that is at most 1 hour old from the DB.
/// - We can get a BitMEX index price for the `end_date` of the [`FundingRate`].
/// - There is no other [`FundingFeeEvent`] in the DB with the same `position_id` and `end_date`.
/// - The position was created _before_ the `end_date` of the [`FundingRate`].
///
/// This function should be safe to retry. Retry should come in handy if the index price is
/// not available.
fn generate_funding_fee_events(
    pool: &Pool<ConnectionManager<PgConnection>>,
    index_price_source: IndexPriceSource,
    auth_users_notifier: tokio::sync::mpsc::Sender<OrderbookMessage>,
) -> Result<()> {
    let mut conn = pool.get()?;

    tracing::debug!("Generating funding fee events");

    let funding_rate = db::funding_rates::get_funding_rate_charged_in_the_last_hour(&mut conn)?;

    let funding_rate = match funding_rate {
        Some(funding_rate) => funding_rate,
        None => {
            tracing::debug!("No current funding rate for this hour");
            return Ok(());
        }
    };

    // TODO: Funding rates should be specific to contract symbols.
    let contract_symbol = ContractSymbol::BtcUsd;

    let index_price = match index_price_source {
        IndexPriceSource::Bitmex => block_in_place(move || {
            let current_index_price =
                get_bitmex_index_price(&contract_symbol, funding_rate.end_date)?;

            anyhow::Ok(current_index_price)
        })?,
        IndexPriceSource::Test => {
            #[cfg(not(debug_assertions))]
            compile_error!("Cannot use a test index price in release mode");

            dec!(50_000)
        }
    };

    if index_price.is_zero() {
        bail!("Cannot generate funding fee events with zero index price");
    }

    // We exclude active positions which were open after this funding period ended.
    let positions = db::positions::Position::get_all_active_positions_open_before(
        &mut conn,
        funding_rate.end_date,
    )?;
    for position in positions {
        let amount = calculate_funding_fee(
            position.quantity,
            funding_rate.rate,
            index_price,
            position.trader_direction,
        );

        if let Some(funding_fee_event) = db::funding_fee_events::insert(
            &mut conn,
            amount,
            position.trader,
            position.id,
            funding_rate.end_date,
            index_price,
            funding_rate.rate,
        )
        .context("Failed to insert funding fee event")?
        {
            block_in_place(|| {
                auth_users_notifier
                    .blocking_send(OrderbookMessage::TraderMessage {
                        trader_id: position.trader,
                        message: Message::FundingFeeEvent(xxi_node::FundingFeeEvent {
                            contract_symbol,
                            contracts: decimal_from_f32(position.quantity),
                            direction: position.trader_direction,
                            price: funding_fee_event.price,
                            fee: funding_fee_event.amount,
                            due_date: funding_fee_event.due_date,
                        }),
                        notification: None,
                    })
                    .map_err(anyhow::Error::new)
                    .context("Could not send pending funding fee event to trader")
            })?;

            tracing::debug!(
                position_id = %position.id,
                trader_pubkey = %position.trader,
                fee_amount = ?amount,
                ?funding_rate,
                "Generated funding fee event"
            );
        }
    }

    anyhow::Ok(())
}

/// Calculate the funding fee.
///
/// We assume that the `index_price` is not zero. Otherwise, the function panics.
fn calculate_funding_fee(
    quantity: f32,
    // Positive means longs pay shorts; negative means shorts pay longs.
    funding_rate: Decimal,
    index_price: Decimal,
    trader_direction: Direction,
) -> SignedAmount {
    // Transform the funding rate from a global perspective (longs and shorts) to a local
    // perspective (the coordinator-trader position).
    let funding_rate = match trader_direction {
        Direction::Long => funding_rate,
        Direction::Short => -funding_rate,
    };

    let quantity = Decimal::try_from(quantity).expect("to fit");

    // E.g. 500 [$] / 20_000 [$/BTC] = 0.025 [BTC]
    let mark_value = quantity / index_price;

    let funding_fee_btc = mark_value * funding_rate;
    let funding_fee_btc = funding_fee_btc
        .round_dp_with_strategy(8, RoundingStrategy::MidpointAwayFromZero)
        .to_f64()
        .expect("to fit");

    SignedAmount::from_btc(funding_fee_btc).expect("to fit")
}

fn get_bitmex_index_price(
    contract_symbol: &ContractSymbol,
    timestamp: OffsetDateTime,
) -> Result<Decimal> {
    let symbol = bitmex_symbol(contract_symbol);

    let time_format = format_description::parse("[year]-[month]-[day] [hour]:[minute]")?;

    // Ideally we get the price indicated by `timestamp`, but if it is not available we are happy to
    // take a price up to 1 minute in the past.
    let start_time = (timestamp - 1.minutes()).format(&time_format)?;
    let end_time = timestamp.format(&time_format)?;

    let mut url = reqwest::Url::parse("https://www.bitmex.com/api/v1/instrument/compositeIndex")?;
    url.query_pairs_mut()
        .append_pair("symbol", &format!(".{symbol}"))
        .append_pair(
            "filter",
            // The `reference` is set to `BMI` to get the _composite_ index.
            &format!("{{\"symbol\": \".{symbol}\", \"startTime\": \"{start_time}\", \"endTime\": \"{end_time}\", \"reference\": \"BMI\"}}"),
        )
        .append_pair("columns", "lastPrice,timestamp,reference")
        // Reversed to get the latest one.
        .append_pair("reverse", "true")
        // Only need one index.
        .append_pair("count", "1");

    let indices = reqwest::blocking::get(url)?.json::<Vec<Index>>()?;
    let index = &indices[0];

    let index_price = Decimal::try_from(index.last_price)?;

    Ok(index_price)
}

fn bitmex_symbol(contract_symbol: &ContractSymbol) -> &str {
    match contract_symbol {
        ContractSymbol::BtcUsd => "BXBT",
    }
}

#[derive(serde::Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
struct Index {
    #[serde(with = "time::serde::rfc3339")]
    #[serde(rename = "timestamp")]
    _timestamp: OffsetDateTime,
    last_price: f64,
    #[serde(rename = "reference")]
    _reference: String,
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use rust_decimal_macros::dec;

    #[test]
    fn calculate_funding_fee_test() {
        assert_debug_snapshot!(calculate_funding_fee(
            500.0,
            dec!(0.003),
            dec!(20_000),
            Direction::Long
        ));
        assert_debug_snapshot!(calculate_funding_fee(
            500.0,
            dec!(0.003),
            dec!(20_000),
            Direction::Short
        ));
        assert_debug_snapshot!(calculate_funding_fee(
            500.0,
            dec!(-0.003),
            dec!(20_000),
            Direction::Long
        ));
        assert_debug_snapshot!(calculate_funding_fee(
            500.0,
            dec!(-0.003),
            dec!(20_000),
            Direction::Short
        ));
        assert_debug_snapshot!(calculate_funding_fee(
            500.0,
            dec!(0.003),
            dec!(40_000),
            Direction::Long
        ));
        assert_debug_snapshot!(calculate_funding_fee(
            500.0,
            dec!(0.003),
            dec!(40_000),
            Direction::Short
        ));
        assert_debug_snapshot!(calculate_funding_fee(
            100.0,
            dec!(0.003),
            dec!(20_000),
            Direction::Long
        ));
        assert_debug_snapshot!(calculate_funding_fee(
            100.0,
            dec!(0.003),
            dec!(20_000),
            Direction::Short
        ));
    }
}
