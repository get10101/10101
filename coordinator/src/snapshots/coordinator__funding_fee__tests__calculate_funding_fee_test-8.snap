---
source: coordinator/src/funding_fee.rs
expression: "calculate_funding_fee(100.0, dec!(0.003), dec!(20_000), Direction::Short)"
---
SignedAmount(-0.00001500 BTC)
